src/System.Management.Automation/engine/lang/parserutils.cs

```diff
@@ -635,19 +635,49 @@ private static object SplitWithPredicate(ExecutionContext context, IScriptExtent
             List<string> results = new List<string>();
             foreach (string item in content)
             {
-                List<string> split = new List<String>();
+                int maxSubstrings = System.Math.Abs(limit);
+                bool rightToLeft = false;
+                if(limit < 0){
+                    // User supplied a negative Max-substrings argument,
+                    // employ Right-to-Left splitting instead
+                    rightToLeft = true;
+                }
 
-                if (limit == 1)
+                if (maxSubstrings == 1)
                 {
                     // Don't bother with looking for any delimiters,
                     // just return the original string.
                     results.Add(item);
                     continue;
                 }
-
+                
+                List<string> split; 
+                if(limit == 0)
+                {
+                    // No limit specified
+                    // set capacity to item.Length + 1
+                    // this covers worst common case (predicate == {$true})
+                    split = new List<String>(item.Length + 1);
+                }
+                else
+                {
+                    // Limit was specified by the user
+                    // instantiate list with maximum needed capacity
+                    split = new List<String>(maxSubstrings);
+                }
+                
                 StringBuilder buf = new StringBuilder();
```

Comment:
We add a char by char in the `buf`. Doesn't make sense to recreate the `buf` after every separator.
Please move out of the cycle to the line 635 `StringBuilder buf = new StringBuilder(256); ` (default 16 is very short for typical strings) and replace the line 691 `buf = new StringBuilder();` on `buf.Clear();`.

Reply:
This is fixed now. `StringBuilder buf` declaration moved outside loop, constructor calls replaced with `.Clear()`